crates/kubelet/src/plugin_watcher/mod.rs

```diff
@@ -0,0 +1,744 @@
+                discovered_path.display(),
+                plugin_info
+            );
+
+            // Step 2: Validate discovered data
+            if let Err(e) = self.validate(&plugin_info, &discovered_path).await {
+                inform_plugin(&discovered_path, Some(e.to_string())).await?;
+                return Err(e).with_context(|| {
+                    format!(
+                        "Validation step failed for plugin discovered at {}",
+                        discovered_path.display()
+                    )
+                });
+            }
+            debug!(
+                "Successfully validated plugin discovered at {}:\n {:?}",
+                discovered_path.display(),
+                plugin_info
+            );
+
+            // Step 3: Register plugin to local storage
+            self.register(&plugin_info, &discovered_path).await;
+
+            // Step 4: Inform plugin
+            inform_plugin(&discovered_path, None).await?;
+            debug!("Plugin registration complete for {:?}", plugin_info)
+        }
+        Ok(())
+    }
+
+    async fn handle_delete(&self, event: Event) {
+        let mut lock = self.plugins.write().await;
+        for deleted_plugin in event
+            .paths
+            .into_iter()
+            .filter(|p| !p.is_dir() && p.extension().unwrap_or_default() == SOCKET_EXTENSION)
+        {
+            let key = match lock.iter().find(|(_, v)| *v.plugin_path == deleted_plugin) {
+                // Take ownership of the key to avoid an immutable borrow
+                Some((key, _)) => key.to_owned(),
+                // If for some reason it is already gone, no need to error
+                None => continue,
+            };
+            lock.remove(&key);
+        }
+    }
+
+    /// Validates the given plugin info gathered from a discovered plugin, returning an error with
+    /// additional information if it is not valid. This will validate 3 specific things (should
+    /// answer YES to all of these):
+    /// 1. Is it a CSIPlugin? If it isn't we will deny it for now. This will be removed as we
+    ///    iterate and if it is needed
+    /// 2. Does the list of supported versions contain the version we expect?
+    /// 3. Is the plugin name available? 3a. If the name is already registered, is the endpoint the
+    ///    exact same? If it is, we allow it to reregister
+    async fn validate(&self, info: &PluginInfo, discovered_path: &PathBuf) -> anyhow::Result<()> {
+        trace!(
+            "Starting validation for plugin {:?} discovered at path {}",
+            info,
+            discovered_path.display()
+        );
+        // Step 1: Check for valid type and if it is a CSIPlugin
+        let plugin_type = PluginType::try_from(info.r#type.as_str())?;
+        trace!("Type validation complete for plugin {:?}", info);
+        if matches!(plugin_type, PluginType::DevicePlugin) {
```

Comment:
Consider allow-listing rather than deny-listing.  Also consider pulling out a helper constant `ALLOWED_PLUGIN_TYPES` and a helper function `is_allowed_plugin_type(&PluginInfo)`.

Reply:
Since this is the only type we have to check for and it may end up being removed if we add support in the future, is there any benefit in doing this? Doesn't feel like it adds much benefit in this case

Reply:
The benefit of allow-listing instead of deny-listing is that we _opt into_ plugin types when we are ready to support them.  The current test says "hey, we support everything except device plugins!"  Whereas actually we only support CSI plugins.  It _so happens_ that you know that these _so happen_ to amount to the same thing.  But I feel that it matters _for the understanding of others_ (and to a lesser extent for future-proofing) how the code expresses it.

The benefit of pulling out a helper function is _naming_.  Don't make me deliver the rant about how functions are primarily there for naming and only incidentally for reuse.  You wouldn't like me when I'm delivering the rant about etc. etc.

Reply:
I still don't think this really adds much in way of clarity given the use of an enum and feels much more like ceremoney, but I'll roll with with it for now as I want to see if I like it as much as you do 😄

Reply:
After implementing it, I personally think that it just adds more code with a zero to possibly negative readability increase. If you feel strongly about it, I'll keep it